File: bms/go/go-gt/updateTrace/updateTrace.go
package test

var ()

const ()

type Vector []int64

func updateTrace(n int64, p, best_p Vector, inc *int64, r int64, trace *Matrix) {
	var i int64
	for i = 0; i < n && p[i] == best_p[i]; i++ { // skip
	}
	if i == n {
		(*inc)++
		initTrace(n, *inc, trace)
	} else {
		for i = 0; i < n; i++ {
			trace.Set(i, p[i], trace.Get(i, p[i])+*inc)
			trace.Set(i, best_p[i], trace.Get(i, best_p[i])+r)
		}
	}
}

func initTrace(n, inc int64, trace *Matrix) {
	var i, j int64
	for i = 0; i < n; i++ {
		for j = 0; j < n; j++ {
			trace.Set(i, j, inc)
		}
	}
}

func (m Matrix) Set(i int64, j int64, v int64) {
	m.A[i*m.N+j] = v
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/fail/go/go-gt/updateTrace/claude2/Hinted/updateTrace.rs

use std::ops::IndexMut;

type Vector = Vec<i64>;

fn update_trace(n: i64, p: &Vector, best_p: &Vector, inc: &mut i64, r: i64, trace: &mut Matrix) {
    let mut i = 0;
    while i < n && p[i as usize] == best_p[i as usize] {
        i += 1;
    }
    if i == n {
        *inc += 1;
        init_trace(n, *inc, trace);
    } else {
        for i in 0..n {
            set(trace, i as usize, p[i as usize] as usize, get(trace, i as usize, p[i as usize] as usize) + *inc);
            set(trace, i as usize, best_p[i as usize] as usize, get(trace, i as usize, best_p[i as usize] as usize) + r);
        }
    }
}

fn init_trace(n: i64, inc: i64, trace: &mut Matrix) {
    for i in 0..n {
        for j in 0..n {
            set(trace, i as usize, j as usize, inc);
        }
    }
}

fn set(m: &mut Matrix, i: usize, j: usize, v: i64) {
    m.data[i][j] = v;
}

fn get(m: &Matrix, i: usize, j: usize) -> i64 {
    m.data[i][j]
}

struct Matrix {
    n: usize,
    data: Vec<Vec<i64>>,
}

